internal/pkg/gateway/api.go

```diff
@@ -427,51 +427,109 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 		return nil, status.Errorf(codes.Unavailable, "no orderer nodes available")
 	}
 
-	// try each orderer in random order
 	var errDetails []proto.Message
-	for _, index := range rand.Perm(len(orderers)) {
-		orderer := orderers[index]
-		logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress)
-
-		var response *ab.BroadcastResponse
-		var err error
-		done := make(chan struct{})
-		go func() {
-			defer close(done)
-			ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
-			defer cancel()
-
-			response, err = gs.broadcast(ctx, orderer, txn)
-		}()
-		select {
-		case <-done:
-			// Broadcast completed normally
-		case <-ctx.Done():
-			// Overall submit timeout expired
-			logger.Warn("Submit call timed out while broadcasting to ordering service")
-			return nil, newRpcError(codes.DeadlineExceeded, "submit timeout expired while broadcasting to ordering service")
+	config := gs.getChannelConfig(request.ChannelId)
+	if config.ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		// For BFT, we send transaction to ALL orderers
+		waitCh := make(chan *gp.ErrorDetail, len(orderers))
+		for _, o := range orderers {
+			go func(ord *orderer) {
+				logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
```

Comment:
I think the way the code is structured now, we have a potential resource accumulation problem.

Imagine you have an orderer that is byzantine and is not responding but you can connect to it. 

You're going to be stuck on: `response, err := broadcast.Recv()` for ever. Because the context's timeout is only on the establishment of the stream, but not on receiving the message of the status back. 

What we need, is to cancel the stream's context when the function that *spawned* the goroutine has ended. Otherwise, this goroutine will live for ever and will never free resources and memory. 

To do that, you need a parent context, that inherits from the client's context, and then the goroutine's context with timeout inherits from the parent context. 

When you exit the function, you cancel the parent context and this way all streams that are "stuck" receiving will be cancelled.